src/include/connection.h

```diff
@@ -347,8 +347,27 @@ struct __wt_connection_impl {
      * array when only a few threads are running.
      */
     WT_SESSION_IMPL *sessions; /* Session reference */
-    uint32_t session_size;     /* Session array size */
-    uint32_t session_cnt;      /* Session count */
+
+#define WT_CONNECTION_FOREACH_SESSION(session, conn)                                             \
+    do {                                                                                         \
+        /*                                                                                       \
+         * No lock is required because the session array is fixed size, but it may contain       \
+         * inactive entries. We must review any active session, so insert a read barrier after   \
+         * reading the active session count. This ensures that the session count is read once at \
```

Comment:
I think this comment is correct but not describing the intent behind the ordered read. On the writer side we have:
```
Initialize session in session array
WRITE_BARRIER
increment session count
```
By reading a session count of N we are guaranteed to see at least N initialized sessions. I think we should update the comment to reflect that.

Reply:
I think there's more than initialised sessions here. In `__wt_session_close_internal` we have the following snippet
```
    while (conn->sessions[conn->session_cnt - 1].active == 0)
        if (--conn->session_cnt == 0)
            break;
```
so without the barrier we could be iterating through the sessions, a session gets closed and `session_cnt` decremented, and we end the loop earlier than we intend should. (Note I didn't see any write barriers around this section of code). 

My read of this `FOREACH` macro is that it ensures that we read all sessions that were active at a specific point in time rather than ensuring we only look at a session post-initialisation.

Reply:
Really because I think its word for word the exact publish case that a WRITE and READ barrier pairing should be used for. The additional while loop in the background seems like an afterthought.

Reply:
Discussed offline. We'll review the usage of `session_cnt` further in WT-11217.